File: ios/native/MauiEmarsys/DotnetEmarsysGeofence.swift
//
//  Copyright © 2024 Emarsys. All rights reserved.
//

import Foundation
import EmarsysSDK

@objc(DotnetEmarsysGeofence)
public class DotnetEmarsysGeofence: NSObject {
    
    @objc
    public static func setInitialEnterTriggerEnabled(_ initialEnterTriggerEnabled: Bool) {
        Emarsys.geofence.initialEnterTriggerEnabled = initialEnterTriggerEnabled
    }
    
    @objc
    public static func enable(_ completionBlock: @escaping CompletionBlock) {
        Emarsys.geofence.enable(completionBlock: completionBlock)
    }
    
    @objc
    public static func disable() {
        Emarsys.geofence.disable()
    }
    
    @objc
    public static func isEnabled() -> Bool {
        return Emarsys.geofence.isEnabled()
    }
    
    @objc
    public static func setEventHandler(_ eventHandler: @escaping EventHandler) {
        Emarsys.geofence.eventHandler = eventHandler
    }
    
    @objc
    public static func registeredGeofences() -> [EMSGeofence] {
        return Emarsys.geofence.registeredGeofences()
    }
    
}

File: android/native/emarsys/src/main/java/com/emarsys/maui/DotnetEmarsysGeofence.java
package com.emarsys.maui;

import androidx.annotation.NonNull;
import com.emarsys.Emarsys;
import com.emarsys.mobileengage.api.geofence.Geofence;
import com.emarsys.mobileengage.api.geofence.Trigger;
import java.util.ArrayList;
import java.util.List;

public class DotnetEmarsysGeofence {

    public static void setInitialEnterTriggerEnabled(boolean initialEnterTriggerEnabled) {
        Emarsys.getGeofence().setInitialEnterTriggerEnabled(initialEnterTriggerEnabled);
    }

    public static void enable(@NonNull CompletionListener completionListener) {
        Emarsys.getGeofence().enable(completionListener::onCompleted);
    }

    public static void disable() {
        Emarsys.getGeofence().disable();
    }

    public static boolean isEnabled() {
        return Emarsys.getGeofence().isEnabled();
    }

    public static void setEventHandler(@NonNull EventHandler eventHandler) {
        Emarsys.getGeofence().setEventHandler(eventHandler::handleEvent);
    }

    public static @NonNull List<EMSGeofence> getRegisteredGeofences() {
        List<Geofence> geofences = Emarsys.getGeofence().getRegisteredGeofences();
        return mapGeofences(geofences);
    }

    static @NonNull List<EMSGeofence> mapGeofences(@NonNull List<Geofence> geofences) {
        List<EMSGeofence> _geofences = new ArrayList<>();
        for (int i = 0; i < geofences.size(); i++) {
            Geofence g = geofences.get(i);
            List<EMSGeofenceTrigger> triggers = new ArrayList<>();
            for (int j = 0; j < g.getTriggers().size(); j++) {
                Trigger t = g.getTriggers().get(j);
                triggers.add(new EMSGeofenceTrigger(t.getId(), t.getType().toString(), t.getLoiteringDelay(), t.getAction()));
            }
            _geofences.add(new EMSGeofence(g.getId(), g.getLat(), g.getLon(), g.getRadius(), g.getWaitInterval(), triggers));
        }
        return _geofences;
    }

}
